// onnxruntime-web@1.22.0-dev.20250409-89f8206ba4/lib/onnxjs/ort-schema/flatbuffers/onnxruntime/fbs.ts
// automatically generated by the FlatBuffers compiler, do not modify

/* eslint-disable @typescript-eslint/no-unused-vars, @typescript-eslint/no-explicit-any, @typescript-eslint/no-non-null-assertion */

export { ArgType } from './fbs/arg-type.js';
export { ArgTypeAndIndex } from './fbs/arg-type-and-index.js';
export { Attribute } from './fbs/attribute.js';
export { AttributeType } from './fbs/attribute-type.js';
export { DeprecatedKernelCreateInfos } from './fbs/deprecated-kernel-create-infos.js';
export { DeprecatedNodeIndexAndKernelDefHash } from './fbs/deprecated-node-index-and-kernel-def-hash.js';
export { DeprecatedSessionState } from './fbs/deprecated-session-state.js';
export { DeprecatedSubGraphSessionState } from './fbs/deprecated-sub-graph-session-state.js';
export { Dimension } from './fbs/dimension.js';
export { DimensionValue } from './fbs/dimension-value.js';
export { DimensionValueType } from './fbs/dimension-value-type.js';
export { EdgeEnd } from './fbs/edge-end.js';
export { Graph } from './fbs/graph.js';
export { InferenceSession } from './fbs/inference-session.js';
export { KernelTypeStrArgsEntry } from './fbs/kernel-type-str-args-entry.js';
export { KernelTypeStrResolver } from './fbs/kernel-type-str-resolver.js';
export { MapType } from './fbs/map-type.js';
export { Model } from './fbs/model.js';
export { Node } from './fbs/node.js';
export { NodeEdge } from './fbs/node-edge.js';
export { NodeType } from './fbs/node-type.js';
export { NodesToOptimizeIndices } from './fbs/nodes-to-optimize-indices.js';
export { OpIdKernelTypeStrArgsEntry } from './fbs/op-id-kernel-type-str-args-entry.js';
export { OperatorSetId } from './fbs/operator-set-id.js';
export { RuntimeOptimizationRecord } from './fbs/runtime-optimization-record.js';
export { RuntimeOptimizationRecordContainerEntry } from './fbs/runtime-optimization-record-container-entry.js';
export { RuntimeOptimizations } from './fbs/runtime-optimizations.js';
export { SequenceType } from './fbs/sequence-type.js';
export { Shape } from './fbs/shape.js';
export { SparseTensor } from './fbs/sparse-tensor.js';
export { StringStringEntry } from './fbs/string-string-entry.js';
export { Tensor } from './fbs/tensor.js';
export { TensorDataType } from './fbs/tensor-data-type.js';
export { TensorTypeAndShape } from './fbs/tensor-type-and-shape.js';
export { TypeInfo } from './fbs/type-info.js';
export { TypeInfoValue } from './fbs/type-info-value.js';
export { ValueInfo } from './fbs/value-info.js';
