// flatbuffers@25.2.10/js/flatbuffers.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ByteBuffer = exports.Builder = exports.Encoding = exports.isLittleEndian = exports.float64 = exports.float32 = exports.int32 = exports.SIZE_PREFIX_LENGTH = exports.FILE_IDENTIFIER_LENGTH = exports.SIZEOF_INT = exports.SIZEOF_SHORT = void 0;
var constants_js_1 = require("./constants.js");
Object.defineProperty(exports, "SIZEOF_SHORT", { enumerable: true, get: function () { return constants_js_1.SIZEOF_SHORT; } });
var constants_js_2 = require("./constants.js");
Object.defineProperty(exports, "SIZEOF_INT", { enumerable: true, get: function () { return constants_js_2.SIZEOF_INT; } });
var constants_js_3 = require("./constants.js");
Object.defineProperty(exports, "FILE_IDENTIFIER_LENGTH", { enumerable: true, get: function () { return constants_js_3.FILE_IDENTIFIER_LENGTH; } });
var constants_js_4 = require("./constants.js");
Object.defineProperty(exports, "SIZE_PREFIX_LENGTH", { enumerable: true, get: function () { return constants_js_4.SIZE_PREFIX_LENGTH; } });
var utils_js_1 = require("./utils.js");
Object.defineProperty(exports, "int32", { enumerable: true, get: function () { return utils_js_1.int32; } });
Object.defineProperty(exports, "float32", { enumerable: true, get: function () { return utils_js_1.float32; } });
Object.defineProperty(exports, "float64", { enumerable: true, get: function () { return utils_js_1.float64; } });
Object.defineProperty(exports, "isLittleEndian", { enumerable: true, get: function () { return utils_js_1.isLittleEndian; } });
var encoding_js_1 = require("./encoding.js");
Object.defineProperty(exports, "Encoding", { enumerable: true, get: function () { return encoding_js_1.Encoding; } });
var builder_js_1 = require("./builder.js");
Object.defineProperty(exports, "Builder", { enumerable: true, get: function () { return builder_js_1.Builder; } });
var byte_buffer_js_1 = require("./byte-buffer.js");
Object.defineProperty(exports, "ByteBuffer", { enumerable: true, get: function () { return byte_buffer_js_1.ByteBuffer; } });
