// onnxruntime-web@1.22.0-dev.20250409-89f8206ba4/lib/onnxjs/ort-schema/flatbuffers/onnxruntime/fbs/attribute-type.js
'use strict';
// automatically generated by the FlatBuffers compiler, do not modify
Object.defineProperty(exports, '__esModule', { value: true });
exports.AttributeType = void 0;
/* eslint-disable @typescript-eslint/no-unused-vars, @typescript-eslint/no-explicit-any, @typescript-eslint/no-non-null-assertion */
var AttributeType;
(function (AttributeType) {
  AttributeType[(AttributeType['UNDEFINED'] = 0)] = 'UNDEFINED';
  AttributeType[(AttributeType['FLOAT'] = 1)] = 'FLOAT';
  AttributeType[(AttributeType['INT'] = 2)] = 'INT';
  AttributeType[(AttributeType['STRING'] = 3)] = 'STRING';
  AttributeType[(AttributeType['TENSOR'] = 4)] = 'TENSOR';
  AttributeType[(AttributeType['GRAPH'] = 5)] = 'GRAPH';
  AttributeType[(AttributeType['FLOATS'] = 6)] = 'FLOATS';
  AttributeType[(AttributeType['INTS'] = 7)] = 'INTS';
  AttributeType[(AttributeType['STRINGS'] = 8)] = 'STRINGS';
  AttributeType[(AttributeType['TENSORS'] = 9)] = 'TENSORS';
  AttributeType[(AttributeType['GRAPHS'] = 10)] = 'GRAPHS';
  AttributeType[(AttributeType['SPARSE_TENSOR'] = 11)] = 'SPARSE_TENSOR';
  AttributeType[(AttributeType['SPARSE_TENSORS'] = 12)] = 'SPARSE_TENSORS';
})(AttributeType || (exports.AttributeType = AttributeType = {}));
