// onnxruntime-web@1.22.0-dev.20250409-89f8206ba4/lib/onnxjs/ort-schema/flatbuffers/onnxruntime/fbs/deprecated-kernel-create-infos.js
'use strict';
// automatically generated by the FlatBuffers compiler, do not modify
var __createBinding =
  (this && this.__createBinding) ||
  (Object.create
    ? function (o, m, k, k2) {
        if (k2 === undefined) k2 = k;
        var desc = Object.getOwnPropertyDescriptor(m, k);
        if (!desc || ('get' in desc ? !m.__esModule : desc.writable || desc.configurable)) {
          desc = {
            enumerable: true,
            get: function () {
              return m[k];
            },
          };
        }
        Object.defineProperty(o, k2, desc);
      }
    : function (o, m, k, k2) {
        if (k2 === undefined) k2 = k;
        o[k2] = m[k];
      });
var __setModuleDefault =
  (this && this.__setModuleDefault) ||
  (Object.create
    ? function (o, v) {
        Object.defineProperty(o, 'default', { enumerable: true, value: v });
      }
    : function (o, v) {
        o['default'] = v;
      });
var __importStar =
  (this && this.__importStar) ||
  function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null)
      for (var k in mod)
        if (k !== 'default' && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
  };
Object.defineProperty(exports, '__esModule', { value: true });
exports.DeprecatedKernelCreateInfos = void 0;
/* eslint-disable @typescript-eslint/no-unused-vars, @typescript-eslint/no-explicit-any, @typescript-eslint/no-non-null-assertion */
const flatbuffers = __importStar(require('flatbuffers'));
/**
 * deprecated: no longer using kernel def hashes
 */
class DeprecatedKernelCreateInfos {
  constructor() {
    this.bb = null;
    this.bb_pos = 0;
  }
  __init(i, bb) {
    this.bb_pos = i;
    this.bb = bb;
    return this;
  }
  static getRootAsDeprecatedKernelCreateInfos(bb, obj) {
    return (obj || new DeprecatedKernelCreateInfos()).__init(bb.readInt32(bb.position()) + bb.position(), bb);
  }
  static getSizePrefixedRootAsDeprecatedKernelCreateInfos(bb, obj) {
    bb.setPosition(bb.position() + flatbuffers.SIZE_PREFIX_LENGTH);
    return (obj || new DeprecatedKernelCreateInfos()).__init(bb.readInt32(bb.position()) + bb.position(), bb);
  }
  nodeIndices(index) {
    const offset = this.bb.__offset(this.bb_pos, 4);
    return offset ? this.bb.readUint32(this.bb.__vector(this.bb_pos + offset) + index * 4) : 0;
  }
  nodeIndicesLength() {
    const offset = this.bb.__offset(this.bb_pos, 4);
    return offset ? this.bb.__vector_len(this.bb_pos + offset) : 0;
  }
  nodeIndicesArray() {
    const offset = this.bb.__offset(this.bb_pos, 4);
    return offset
      ? new Uint32Array(
          this.bb.bytes().buffer,
          this.bb.bytes().byteOffset + this.bb.__vector(this.bb_pos + offset),
          this.bb.__vector_len(this.bb_pos + offset),
        )
      : null;
  }
  kernelDefHashes(index) {
    const offset = this.bb.__offset(this.bb_pos, 6);
    return offset ? this.bb.readUint64(this.bb.__vector(this.bb_pos + offset) + index * 8) : BigInt(0);
  }
  kernelDefHashesLength() {
    const offset = this.bb.__offset(this.bb_pos, 6);
    return offset ? this.bb.__vector_len(this.bb_pos + offset) : 0;
  }
  static startDeprecatedKernelCreateInfos(builder) {
    builder.startObject(2);
  }
  static addNodeIndices(builder, nodeIndicesOffset) {
    builder.addFieldOffset(0, nodeIndicesOffset, 0);
  }
  static createNodeIndicesVector(builder, data) {
    builder.startVector(4, data.length, 4);
    for (let i = data.length - 1; i >= 0; i--) {
      builder.addInt32(data[i]);
    }
    return builder.endVector();
  }
  static startNodeIndicesVector(builder, numElems) {
    builder.startVector(4, numElems, 4);
  }
  static addKernelDefHashes(builder, kernelDefHashesOffset) {
    builder.addFieldOffset(1, kernelDefHashesOffset, 0);
  }
  static createKernelDefHashesVector(builder, data) {
    builder.startVector(8, data.length, 8);
    for (let i = data.length - 1; i >= 0; i--) {
      builder.addInt64(data[i]);
    }
    return builder.endVector();
  }
  static startKernelDefHashesVector(builder, numElems) {
    builder.startVector(8, numElems, 8);
  }
  static endDeprecatedKernelCreateInfos(builder) {
    const offset = builder.endObject();
    return offset;
  }
  static createDeprecatedKernelCreateInfos(builder, nodeIndicesOffset, kernelDefHashesOffset) {
    DeprecatedKernelCreateInfos.startDeprecatedKernelCreateInfos(builder);
    DeprecatedKernelCreateInfos.addNodeIndices(builder, nodeIndicesOffset);
    DeprecatedKernelCreateInfos.addKernelDefHashes(builder, kernelDefHashesOffset);
    return DeprecatedKernelCreateInfos.endDeprecatedKernelCreateInfos(builder);
  }
}
exports.DeprecatedKernelCreateInfos = DeprecatedKernelCreateInfos;
