// onnxruntime-web@1.22.0-dev.20250409-89f8206ba4/lib/onnxjs/ort-schema/flatbuffers/onnxruntime/fbs/deprecated-node-index-and-kernel-def-hash.js
'use strict';
// automatically generated by the FlatBuffers compiler, do not modify
var __createBinding =
  (this && this.__createBinding) ||
  (Object.create
    ? function (o, m, k, k2) {
        if (k2 === undefined) k2 = k;
        var desc = Object.getOwnPropertyDescriptor(m, k);
        if (!desc || ('get' in desc ? !m.__esModule : desc.writable || desc.configurable)) {
          desc = {
            enumerable: true,
            get: function () {
              return m[k];
            },
          };
        }
        Object.defineProperty(o, k2, desc);
      }
    : function (o, m, k, k2) {
        if (k2 === undefined) k2 = k;
        o[k2] = m[k];
      });
var __setModuleDefault =
  (this && this.__setModuleDefault) ||
  (Object.create
    ? function (o, v) {
        Object.defineProperty(o, 'default', { enumerable: true, value: v });
      }
    : function (o, v) {
        o['default'] = v;
      });
var __importStar =
  (this && this.__importStar) ||
  function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null)
      for (var k in mod)
        if (k !== 'default' && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
  };
Object.defineProperty(exports, '__esModule', { value: true });
exports.DeprecatedNodeIndexAndKernelDefHash = void 0;
/* eslint-disable @typescript-eslint/no-unused-vars, @typescript-eslint/no-explicit-any, @typescript-eslint/no-non-null-assertion */
const flatbuffers = __importStar(require('flatbuffers'));
/**
 * deprecated: no longer using kernel def hashes
 */
class DeprecatedNodeIndexAndKernelDefHash {
  constructor() {
    this.bb = null;
    this.bb_pos = 0;
  }
  __init(i, bb) {
    this.bb_pos = i;
    this.bb = bb;
    return this;
  }
  static getRootAsDeprecatedNodeIndexAndKernelDefHash(bb, obj) {
    return (obj || new DeprecatedNodeIndexAndKernelDefHash()).__init(bb.readInt32(bb.position()) + bb.position(), bb);
  }
  static getSizePrefixedRootAsDeprecatedNodeIndexAndKernelDefHash(bb, obj) {
    bb.setPosition(bb.position() + flatbuffers.SIZE_PREFIX_LENGTH);
    return (obj || new DeprecatedNodeIndexAndKernelDefHash()).__init(bb.readInt32(bb.position()) + bb.position(), bb);
  }
  nodeIndex() {
    const offset = this.bb.__offset(this.bb_pos, 4);
    return offset ? this.bb.readUint32(this.bb_pos + offset) : 0;
  }
  kernelDefHash() {
    const offset = this.bb.__offset(this.bb_pos, 6);
    return offset ? this.bb.readUint64(this.bb_pos + offset) : BigInt('0');
  }
  static startDeprecatedNodeIndexAndKernelDefHash(builder) {
    builder.startObject(2);
  }
  static addNodeIndex(builder, nodeIndex) {
    builder.addFieldInt32(0, nodeIndex, 0);
  }
  static addKernelDefHash(builder, kernelDefHash) {
    builder.addFieldInt64(1, kernelDefHash, BigInt('0'));
  }
  static endDeprecatedNodeIndexAndKernelDefHash(builder) {
    const offset = builder.endObject();
    return offset;
  }
  static createDeprecatedNodeIndexAndKernelDefHash(builder, nodeIndex, kernelDefHash) {
    DeprecatedNodeIndexAndKernelDefHash.startDeprecatedNodeIndexAndKernelDefHash(builder);
    DeprecatedNodeIndexAndKernelDefHash.addNodeIndex(builder, nodeIndex);
    DeprecatedNodeIndexAndKernelDefHash.addKernelDefHash(builder, kernelDefHash);
    return DeprecatedNodeIndexAndKernelDefHash.endDeprecatedNodeIndexAndKernelDefHash(builder);
  }
}
exports.DeprecatedNodeIndexAndKernelDefHash = DeprecatedNodeIndexAndKernelDefHash;
