// onnxruntime-web@1.22.0-dev.20250409-89f8206ba4/lib/onnxjs/ort-schema/flatbuffers/onnxruntime/fbs/kernel-type-str-args-entry.js
'use strict';
// automatically generated by the FlatBuffers compiler, do not modify
var __createBinding =
  (this && this.__createBinding) ||
  (Object.create
    ? function (o, m, k, k2) {
        if (k2 === undefined) k2 = k;
        var desc = Object.getOwnPropertyDescriptor(m, k);
        if (!desc || ('get' in desc ? !m.__esModule : desc.writable || desc.configurable)) {
          desc = {
            enumerable: true,
            get: function () {
              return m[k];
            },
          };
        }
        Object.defineProperty(o, k2, desc);
      }
    : function (o, m, k, k2) {
        if (k2 === undefined) k2 = k;
        o[k2] = m[k];
      });
var __setModuleDefault =
  (this && this.__setModuleDefault) ||
  (Object.create
    ? function (o, v) {
        Object.defineProperty(o, 'default', { enumerable: true, value: v });
      }
    : function (o, v) {
        o['default'] = v;
      });
var __importStar =
  (this && this.__importStar) ||
  function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null)
      for (var k in mod)
        if (k !== 'default' && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
  };
Object.defineProperty(exports, '__esModule', { value: true });
exports.KernelTypeStrArgsEntry = void 0;
/* eslint-disable @typescript-eslint/no-unused-vars, @typescript-eslint/no-explicit-any, @typescript-eslint/no-non-null-assertion */
const flatbuffers = __importStar(require('flatbuffers'));
const arg_type_and_index_js_1 = require('../../onnxruntime/fbs/arg-type-and-index.js');
class KernelTypeStrArgsEntry {
  constructor() {
    this.bb = null;
    this.bb_pos = 0;
  }
  __init(i, bb) {
    this.bb_pos = i;
    this.bb = bb;
    return this;
  }
  static getRootAsKernelTypeStrArgsEntry(bb, obj) {
    return (obj || new KernelTypeStrArgsEntry()).__init(bb.readInt32(bb.position()) + bb.position(), bb);
  }
  static getSizePrefixedRootAsKernelTypeStrArgsEntry(bb, obj) {
    bb.setPosition(bb.position() + flatbuffers.SIZE_PREFIX_LENGTH);
    return (obj || new KernelTypeStrArgsEntry()).__init(bb.readInt32(bb.position()) + bb.position(), bb);
  }
  kernelTypeStr(optionalEncoding) {
    const offset = this.bb.__offset(this.bb_pos, 4);
    return offset ? this.bb.__string(this.bb_pos + offset, optionalEncoding) : null;
  }
  args(index, obj) {
    const offset = this.bb.__offset(this.bb_pos, 6);
    return offset
      ? (obj || new arg_type_and_index_js_1.ArgTypeAndIndex()).__init(
          this.bb.__indirect(this.bb.__vector(this.bb_pos + offset) + index * 4),
          this.bb,
        )
      : null;
  }
  argsLength() {
    const offset = this.bb.__offset(this.bb_pos, 6);
    return offset ? this.bb.__vector_len(this.bb_pos + offset) : 0;
  }
  static startKernelTypeStrArgsEntry(builder) {
    builder.startObject(2);
  }
  static addKernelTypeStr(builder, kernelTypeStrOffset) {
    builder.addFieldOffset(0, kernelTypeStrOffset, 0);
  }
  static addArgs(builder, argsOffset) {
    builder.addFieldOffset(1, argsOffset, 0);
  }
  static createArgsVector(builder, data) {
    builder.startVector(4, data.length, 4);
    for (let i = data.length - 1; i >= 0; i--) {
      builder.addOffset(data[i]);
    }
    return builder.endVector();
  }
  static startArgsVector(builder, numElems) {
    builder.startVector(4, numElems, 4);
  }
  static endKernelTypeStrArgsEntry(builder) {
    const offset = builder.endObject();
    builder.requiredField(offset, 4); // kernel_type_str
    return offset;
  }
  static createKernelTypeStrArgsEntry(builder, kernelTypeStrOffset, argsOffset) {
    KernelTypeStrArgsEntry.startKernelTypeStrArgsEntry(builder);
    KernelTypeStrArgsEntry.addKernelTypeStr(builder, kernelTypeStrOffset);
    KernelTypeStrArgsEntry.addArgs(builder, argsOffset);
    return KernelTypeStrArgsEntry.endKernelTypeStrArgsEntry(builder);
  }
}
exports.KernelTypeStrArgsEntry = KernelTypeStrArgsEntry;
