// together-ai@0.16.0/resources/endpoints.mjs
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
import { APIResource } from "../resource.mjs";
import { isRequestOptions } from "../core.mjs";
export class Endpoints extends APIResource {
    /**
     * Creates a new dedicated endpoint for serving models. The endpoint will
     * automatically start after creation. You can deploy any supported model on
     * hardware configurations that meet the model's requirements.
     */
    create(body, options) {
        return this._client.post('/endpoints', { body, ...options });
    }
    /**
     * Retrieves details about a specific endpoint, including its current state,
     * configuration, and scaling settings.
     */
    retrieve(endpointId, options) {
        return this._client.get(`/endpoints/${endpointId}`, options);
    }
    /**
     * Updates an existing endpoint's configuration. You can modify the display name,
     * autoscaling settings, or change the endpoint's state (start/stop).
     */
    update(endpointId, body, options) {
        return this._client.patch(`/endpoints/${endpointId}`, { body, ...options });
    }
    list(query = {}, options) {
        if (isRequestOptions(query)) {
            return this.list({}, query);
        }
        return this._client.get('/endpoints', { query, ...options });
    }
    /**
     * Permanently deletes an endpoint. This action cannot be undone.
     */
    delete(endpointId, options) {
        return this._client.delete(`/endpoints/${endpointId}`, {
            ...options,
            headers: { Accept: '*/*', ...options?.headers },
        });
    }
}
